File: src/utils/exactPivotEngine.js
import { compareDateStrings } from './excelParser';

/**
 * Computes exact Pivot Table for NMPG, NOINV, KDPRC, USID with Count JMDOS and Sum NETTO
 * matching the specification from "Contoh tabel.xlsx"
 */
export function buildExactPivot({
  rows = [],
  selectedDate = '',      // Filter TGBON (or 'ALL')
  selectedUser = '',      // Filter USID (or 'ALL')
  selectedJenisB = '',    // Filter JENIS_B (or 'ALL')
  searchQuery = ''
}) {
  if (!rows || rows.length === 0) {
    return {
      availableDates: [],
      availableUsers: [],
      availableJenisB: [],
      filteredGroups: [],
      totalNota: 0,
      totalCountJMDOS: 0,
      totalSumNetto: 0,
      userStats: {},
      dateStats: {},
      grandTotalNotaKeseluruhan: 0,
      grandTotalNettoKeseluruhan: 0,
    };
  }

  // Collect unique filter options from entire dataset
  const dateMap = new Map(); // date -> count distinct NOINV
  const userMap = new Map(); // user -> count distinct NOINV
  const jenisBSet = new Set();
  const allInvoicesSet = new Set();
  let grandTotalNettoAll = 0;

  // Pre-scan for all options
  rows.forEach((r) => {
    const d = String(r.TGBON || r.Tanggal || '').trim();
    const u = String(r.USID || r.User || '').trim();
    const jb = String(r.JENIS_B || r.JENIS || '').trim();
    const noinv = String(r.NOINV || r.noinv || '').trim();
    const netto = Number(String(r.NETTO || 0).replace(/[^0-9.-]+/g, '')) || 0;

    if (d) {
      if (!dateMap.has(d)) dateMap.set(d, new Set());
      if (noinv) dateMap.get(d).add(noinv);
    }
    if (u) {
      if (!userMap.has(u)) userMap.set(u, new Set());
      if (noinv) userMap.get(u).add(noinv);
    }
    if (jb) jenisBSet.add(jb);
    if (noinv) allInvoicesSet.add(noinv);
    grandTotalNettoAll += netto;
  });

  // Sort available dates chronologically (DD-MM-YYYY or YYYY-MM-DD)
  const availableDates = Array.from(dateMap.keys()).sort(compareDateStrings);

  // Sort available users by total invoices
  const availableUsers = Array.from(userMap.keys()).sort((a, b) => {
    return (userMap.get(b)?.size || 0) - (userMap.get(a)?.size || 0);
  });

  const availableJenisB = Array.from(jenisBSet).sort();

  // Apply filters: TGBON, USID, JENIS_B, and searchQuery
  const filteredRows = rows.filter((r) => {
    const d = String(r.TGBON || r.Tanggal || '').trim();
    const u = String(r.USID || r.User || '').trim();
    const jb = String(r.JENIS_B || r.JENIS || '').trim();

    if (selectedDate && selectedDate !== 'ALL' && d !== selectedDate) {
      return false;
    }
    if (selectedUser && selectedUser !== 'ALL' && u !== selectedUser) {
      return false;
    }
    if (selectedJenisB && selectedJenisB !== 'ALL' && jb !== selectedJenisB) {
      return false;
    }

    if (searchQuery) {
      const q = searchQuery.toLowerCase();
      const nmpg = String(r.NMPG || '').toLowerCase();
      const noinv = String(r.NOINV || '').toLowerCase();
      const kdprc = String(r.KDPRC || '').toLowerCase();
      const usid = String(r.USID || '').toLowerCase();
      if (!nmpg.includes(q) && !noinv.includes(q) && !kdprc.includes(q) && !usid.includes(q)) {
        return false;
      }
    }

    return true;
  });

  // Group by (NMPG, NOINV, KDPRC, USID)
  const groupMap = new Map();
  const filteredInvoicesSet = new Set();
  let totalCountJMDOS = 0;
  let totalSumNetto = 0;

  filteredRows.forEach((r) => {
    const nmpg = String(r.NMPG || '(Tanpa Sales)').trim();
    const noinv = String(r.NOINV || '(Tanpa NoInv)').trim();
    const kdprc = String(r.KDPRC || '-').trim();
    const usid = String(r.USID || '-').trim();
    const tgbon = String(r.TGBON || '').trim();
    const key = `${nmpg}|||${noinv}|||${kdprc}|||${usid}`;

    if (!groupMap.has(key)) {
      groupMap.set(key, {
        nmpg,
        noinv,
        kdprc,
        usid,
        tgbon,
        countJMDOS: 0,
        sumNetto: 0,
        rawRows: []
      });
    }

    const g = groupMap.get(key);
    g.countJMDOS += 1;
    const nettoVal = Number(String(r.NETTO || 0).replace(/[^0-9.-]+/g, '')) || 0;
    g.sumNetto += nettoVal;
    g.rawRows.push(r);

    totalCountJMDOS += 1;
    totalSumNetto += nettoVal;
    filteredInvoicesSet.add(noinv);
  });

  // Convert map to array and sort by NMPG, then NOINV
  const filteredGroups = Array.from(groupMap.values()).sort((a, b) => {
    const compNmpg = a.nmpg.localeCompare(b.nmpg);
    if (compNmpg !== 0) return compNmpg;
    return a.noinv.localeCompare(b.noinv);
  });

  // User Stats (Total Nota per User in current filtered context)
  const userStats = {};
  userMap.forEach((invSet, u) => {
    userStats[u] = invSet.size;
  });

  const dateStats = {};
  dateMap.forEach((invSet, d) => {
    dateStats[d] = invSet.size;
  });

  return {
    availableDates,
    availableUsers,
    availableJenisB,
    filteredGroups,
    totalNota: filteredGroups.length, // Each grouped row is 1 distinct NOINV nota
    totalCountJMDOS,
    totalSumNetto,
    userStats,
    dateStats,
    grandTotalNotaKeseluruhan: allInvoicesSet.size,
    grandTotalNettoKeseluruhan: grandTotalNettoAll
  };
}

File: LaporanProduktivitasWPF/Services/ExactPivotEngine.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using LaporanProduktivitasWPF.Models;

namespace LaporanProduktivitasWPF.Services
{
    public class ExactPivotResult
    {
        public List<ExactPivotItem> Items { get; set; }
        public List<string> AvailableDates { get; set; }
        public List<string> AvailableUsers { get; set; }
        public List<string> AvailableJenisB { get; set; }
        public int TotalNota { get; set; }
        public int TotalCountJMDOS { get; set; }
        public double TotalSumNetto { get; set; }
        public Dictionary<string, int> UserStats { get; set; }
        public Dictionary<string, int> DateStats { get; set; }
        public int GrandTotalNotaKeseluruhan { get; set; }
        public double GrandTotalNettoKeseluruhan { get; set; }

        public ExactPivotResult()
        {
            Items = new List<ExactPivotItem>();
            AvailableDates = new List<string>();
            AvailableUsers = new List<string>();
            AvailableJenisB = new List<string>();
            UserStats = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
            DateStats = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        }
    }

    public static class ExactPivotEngine
    {
        public static ExactPivotResult BuildExactPivot(
            IEnumerable<RawRow> rows,
            string selectedDate = "ALL",
            string selectedUser = "ALL",
            string selectedJenisB = "ALL",
            string searchQuery = "")
        {
            var result = new ExactPivotResult();
            if (rows == null) return result;

            var dateMap = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
            var userMap = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
            var jenisBSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var allInvoicesSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            double grandTotalNettoAll = 0.0;

            var rowList = rows.ToList();

            // Pre-scan for all options and grand totals
            foreach (var r in rowList)
            {
                string d = r.Get("TGBON");
                if (string.IsNullOrEmpty(d)) d = r.Get("Tanggal");
                string u = r.Get("USID");
                if (string.IsNullOrEmpty(u)) u = r.Get("User");
                string jb = r.Get("JENIS_B");
                if (string.IsNullOrEmpty(jb)) jb = r.Get("JENIS");
                string noinv = r.Get("NOINV");
                if (string.IsNullOrEmpty(noinv)) noinv = r.Get("No Invoice");

                double netto = ParseDouble(r.Get("NETTO"));

                if (!string.IsNullOrEmpty(d))
                {
                    if (!dateMap.ContainsKey(d)) dateMap[d] = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                    if (!string.IsNullOrEmpty(noinv)) dateMap[d].Add(noinv);
                }

                if (!string.IsNullOrEmpty(u))
                {
                    if (!userMap.ContainsKey(u)) userMap[u] = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                    if (!string.IsNullOrEmpty(noinv)) userMap[u].Add(noinv);
                }

                if (!string.IsNullOrEmpty(jb)) jenisBSet.Add(jb);
                if (!string.IsNullOrEmpty(noinv)) allInvoicesSet.Add(noinv);
                grandTotalNettoAll += netto;
            }

            // Sort dates
            result.AvailableDates = dateMap.Keys.OrderBy(d => d).ToList();
            // Sort users by invoice count desc
            result.AvailableUsers = userMap.OrderByDescending(kv => kv.Value.Count).Select(kv => kv.Key).ToList();
            result.AvailableJenisB = jenisBSet.OrderBy(j => j).ToList();
            result.GrandTotalNotaKeseluruhan = allInvoicesSet.Count;
            result.GrandTotalNettoKeseluruhan = grandTotalNettoAll;

            foreach (var kv in userMap) result.UserStats[kv.Key] = kv.Value.Count;
            foreach (var kv in dateMap) result.DateStats[kv.Key] = kv.Value.Count;

            // Filter rows
            string q = (searchQuery ?? "").Trim().ToLowerInvariant();

            var filtered = rowList.Where(r =>
            {
                string d = r.Get("TGBON");
                if (string.IsNullOrEmpty(d)) d = r.Get("Tanggal");
                string u = r.Get("USID");
                if (string.IsNullOrEmpty(u)) u = r.Get("User");
                string jb = r.Get("JENIS_B");
                if (string.IsNullOrEmpty(jb)) jb = r.Get("JENIS");

                if (!string.IsNullOrEmpty(selectedDate) && selectedDate != "ALL" && !string.Equals(d, selectedDate, StringComparison.OrdinalIgnoreCase))
                    return false;

                if (!string.IsNullOrEmpty(selectedUser) && selectedUser != "ALL" && !string.Equals(u, selectedUser, StringComparison.OrdinalIgnoreCase))
                    return false;

                if (!string.IsNullOrEmpty(selectedJenisB) && selectedJenisB != "ALL" && !string.Equals(jb, selectedJenisB, StringComparison.OrdinalIgnoreCase))
                    return false;

                if (!string.IsNullOrEmpty(q))
                {
                    string nmpg = r.Get("NMPG").ToLowerInvariant();
                    string noinv = r.Get("NOINV").ToLowerInvariant();
                    string kdprc = r.Get("KDPRC").ToLowerInvariant();
                    string usid = r.Get("USID").ToLowerInvariant();
                    if (!nmpg.Contains(q) && !noinv.Contains(q) && !kdprc.Contains(q) && !usid.Contains(q))
                        return false;
                }

                return true;
            });

            // Group by (NMPG, NOINV, KDPRC, USID)
            var groupMap = new Dictionary<string, ExactPivotItem>(StringComparer.OrdinalIgnoreCase);

            foreach (var r in filtered)
            {
                string nmpg = r.Get("NMPG");
                if (string.IsNullOrEmpty(nmpg)) nmpg = "(Tanpa Sales)";
                string noinv = r.Get("NOINV");
                if (string.IsNullOrEmpty(noinv)) noinv = "(Tanpa NoInv)";
                string kdprc = r.Get("KDPRC");
                if (string.IsNullOrEmpty(kdprc)) kdprc = "-";
                string usid = r.Get("USID");
                if (string.IsNullOrEmpty(usid)) usid = "-";
                string tgbon = r.Get("TGBON");
                string jb = r.Get("JENIS_B");

                string key = nmpg + "|||" + noinv + "|||" + kdprc + "|||" + usid;

                ExactPivotItem item;
                if (!groupMap.TryGetValue(key, out item))
                {
                    item = new ExactPivotItem
                    {
                        NMPG = nmpg,
                        NOINV = noinv,
                        KDPRC = kdprc,
                        USID = usid,
                        TGBON = tgbon,
                        JENIS_B = jb,
                        CountJMDOS = 0,
                        SumNetto = 0.0
                    };
                    groupMap[key] = item;
                }

                item.CountJMDOS += 1;
                double nettoVal = ParseDouble(r.Get("NETTO"));
                item.SumNetto += nettoVal;

                result.TotalCountJMDOS += 1;
                result.TotalSumNetto += nettoVal;
            }

            var sortedItems = groupMap.Values
                .OrderBy(item => item.NMPG)
                .ThenBy(item => item.NOINV)
                .ToList();

            for (int i = 0; i < sortedItems.Count; i++)
            {
                sortedItems[i].No = i + 1;
            }

            result.Items = sortedItems;
            result.TotalNota = sortedItems.Count;

            return result;
        }

        private static double ParseDouble(string val)
        {
            if (string.IsNullOrEmpty(val)) return 0.0;
            double d;
            if (double.TryParse(val, NumberStyles.Any, CultureInfo.InvariantCulture, out d)) return d;
            if (double.TryParse(val, NumberStyles.Any, new CultureInfo("id-ID"), out d)) return d;
            string clean = val.Replace(",", ".");
            if (double.TryParse(clean, NumberStyles.Any, CultureInfo.InvariantCulture, out d)) return d;
            return 0.0;
        }
    }
}
